### DISCUSSION TITLE
Rework of chat app

### DISCUSSION THREAD

**Discussion by calderonsamuel** [MichelNivard/gptstudio#77] [Show and tell]
For the past few days I have been working on trying to make the chat app look more like an actual chat window. You can try it with:

```r
remotes::install_github("calderonsamuel/gptstudio")
```

Feel free to test it an let me know if it breaks. Also, I haven't done a PR yet because no one really asked for this.

Here I'll explain incrementally the new features and the reworks.

## New look of the message history

Each individual message is now rounded and has an icon indicating whether it comes from the user or from the assistant. Each role has a different  horizontal aligment and a slightly different background color.

![image](https://user-images.githubusercontent.com/19418298/233134945-06311099-92c0-4f4f-b728-66eb37f67836.png)

## Simplified user inputs

The prompt and buttons have been simplified to give the chat more room to expand.

![image](https://user-images.githubusercontent.com/19418298/233137057-7d0991d8-ab56-4b7f-ae93-e88cba41e600.png)

Now the app has a settings button where the user can still choose its skill level and prefered style.

![image](https://user-images.githubusercontent.com/19418298/233137374-4593410a-3132-4c1a-a886-5fd4966cb7e5.png)

## Welcome message with instructions

When the app starts (or history is cleared) the assistant greets the user with a random welcome message and instructions on how to use the app.

![image](https://user-images.githubusercontent.com/19418298/233138306-675e8693-e44a-4266-a293-070460e39e36.png)

## The chat can be adjusted vertically, horizontally and is scrollable

Limited to 800px width. The prompt input is always fixed to the bottom of the app.

https://user-images.githubusercontent.com/19418298/233140923-5787ee5e-1042-4e84-8a42-6f1a55a47801.mp4

## The chat inherits the current rstudio theme

This makes it look more integrated with the IDE. I believe this gives the feel of what an extension does in VScode.

https://user-images.githubusercontent.com/19418298/233145316-5efe0e77-2192-48e6-97a1-02d87bd37255.mp4

# Notes

Although the main focus of the rework focused on the chat app, some changes in internals were needed. I also had to structure the app in modules to better isolate all internal logic and functions. 

I have documented almost every new function and module that I have added and made sure that the changes pass all the previous tests and the ones that I added.

# Pending features

In my fork issues panel <https://github.com/calderonsamuel/gptstudio/issues> you can see the next features I would like to include (also bugs). In the near future, I'll be working on adding:

- [ ] A copy to clipboard button in code chunks
- [ ] A nicer looking scrollbar 

In the not so near future, I would love to have in the app:

- [ ] Typewriting effect for new messages
- [ ] Custom system prompts
- [ ] Import and export chat history
- [ ] User prompt templates
- [ ] Insert into current document button for code chunks
- [ ] Model selection setting

Please try it and let me know what you think!

**Comment by JamesHWade** [comment]
This is fantastic! I would very much welcome a PR. Great work!!!

**Comment by MichelNivard** [comment]
Amazing! yes please submit a PR would love this!

**Comment by calderonsamuel** [comment]
I will close this discussion because the main changes were implemented. The features for the "not so near future" will be listed in the issues of this repo.
